source/simulator/assembly.cc

```diff
@@ -1146,14 +1151,18 @@ namespace aspect
               // We only need to look up values of shape functions if they
               // belong to 'our' component. They are zero otherwise anyway.
               // Note that we later only look at the values that we do set here.
-              for (unsigned int k=0; k<advection_dofs_per_cell; ++k)
+              for (unsigned int k=0, i=0; i<advection_dofs_per_cell; ++k)
```

Comment:
I agree it is hard to read, but I do not feel your solution makes it much better. The loop head becomes pretty complicated, and we have to evaluate the condition twice. What about I move the increment of k also to the end of the loop, close to the increment of i? That way it is clear, what is increased in which cases, we only have to evaluate once, and the loop head is simple to read.

Also is there a special reason you would like the loop to go over `dofs_per_cell`? I can see that `i<advection_dofs_per_cell` can cause an endless loop, if the condition in the loop is wrong, but apart from that I would favor that one. On the other hand the evaluation of the condition is not too expensive, so if you feel strongly about it I am fine changing the loop back to `k<dofs_per_cell`.